File: Firmware/comm_protocol/lib/Synapse/Synapse.cpp
#include "Arduino.h"
#include "Synapse.h"

static const uint8_t header = 0xAA;
static const uint8_t footer = 0xBB;
static const uint8_t finger_address[6] = {0x01,0x02,0x03,0x04,0x05,0x06};
static const unsigned int dataSize = 4;
static const unsigned int messageSize = 30;

typedef union {
    float floating;
    uint8_t binary[dataSize];
} binaryFloat;

uint8_t check = 0x00;

bool Synapse::waitHeader() {
    uint8_t inByte;
    while(Serial.available()>0) {
        inByte = Serial.read();
        // delayMicroseconds(50);
        if(inByte==header) {
            return true;
        }
        else return false;
    }
    return false;
}

// bool checkMsg(uint8_t *buff) {
//     uint8_t xorChk = 0x00;
//     unsigned int i;
//     for(i=0;i<4;i++) {
//         xorChk = xorChk^buff[i];
//     }
//     if(xorChk==check) return true;
//     else return false;
// }

uint8_t* Synapse::getMessage() {
    static uint8_t message[messageSize];
    uint8_t inByte;
    unsigned int i=0;
    while(!waitHeader()) {}
    while(Serial.available()>0) {
        inByte = Serial.read();
        // delayMicroseconds(50);
        if((inByte!=footer)&&(i<messageSize)) {
            message[i] = inByte;
            i++;
        }
        else if(inByte==footer) {
            return message;
        }
        else return NULL;
    }
    return NULL;
}

float* Synapse::getSetPoints() {
    uint8_t* message;
    binaryFloat data;
    static float setPointArray[6];
    unsigned int i=0, j, k;
    message = getMessage();
    for(j=0;j<6;j++) {
        if(message[i]==finger_address[j]) {
            i++;
            for(k=0;k<dataSize;k++) {
                data.binary[k] = message[i];
                i++;
            }
            setPointArray[j] = data.floating;
        }
    }
    return setPointArray;
}

void Synapse::write(float* messageToSend) {
    binaryFloat data;
    uint8_t message[32];
    message[0] = header;
    unsigned int i=1, j, k;
    for(j=0;j<6;j++) {
        data.floating = messageToSend[j];
        message[i] = finger_address[j];
        i++;
        for(k=0;k<dataSize;k++) {
            message[i] = data.binary[k];
            i++;
        }
    }
    message[31] = footer;
    Serial.write(message,32);
}

// uint8_t* Synapse::getData() {
//     static uint8_t inBuff[dataSize];
//     while(Serial.available()>0) {
//         for(int i = 0;i<dataSize;i++) {
//             inBuff[i] = Serial.read();
//             delayMicroseconds(50);
//         }
//     }
//     return inBuff;
// }

// float* Synapse::getSetPoints() {
//     static float setPointArray[6];
//     uint8_t* inBuff;
//     uint8_t inByte;
//     binaryFloat data;
//     while(!waitHeader()) {}
//     for(int i=0;i<6;i++) {
//         inByte = Serial.read();
//         delayMicroseconds(50);
//         if(inByte == finger_address[i]) {
//             inBuff = getData();
//             for(int j=0;j<dataSize;j++) {
//                 data.binary[j] = inBuff[j];
//             }
//             setPointArray[i] = data.floating;
//         }
//         else if(inByte == footer) {
//             return setPointArray;
//         }
//     }
// }

File: Control/synapse.py
import serial
import time
import struct

HEADER = chr(0x7E)
FOOTER = chr(0x7E)
FINGER_ADDRESS = [chr(0x01),chr(0x02),chr(0x03),chr(0x04),chr(0x05),chr(0x06)]

def scan_ports():
    serial_port_list = []
    portname_start = ["/dev/ttyUSB","/dev/ttyACM","COM"]
    for port in portname_start:
        for i in range(10):
            portnum = str(i)
            portname_full = ''.join([port,portnum])
            try:
                ser = serial.Serial(portname_full, 115200, timeout=1)
                ser.close()
                serial_port_list.append(portname_full)
            except serial.SerialException:
                pass
    return serial_port_list

def connect(port_name):
    try:
        ser = serial.Serial(port_name, 115200, timeout=1)
        ser.setDTR(False)
        time.sleep(1)
        ser.flushInput()
        ser.setDTR(True)
        time.sleep(2)
        status = True
    except serial.SerialException:
        ser = None
        status = False
    return ser, status

def send_setpoint_list(ser,setpoint_list):
    message = HEADER
    for i in range(len(setpoint_list)):
        data = struct.pack('f',float(setpoint_list[i]))
        message += FINGER_ADDRESS[i] + data
    message += FOOTER
    try:
        ser.write(message)
    except serial.SerialException:
        raise

def get_message():
    message = ''
    in_byte = chr(0x00)
    message_size = 30
    i = 0
    while(ser.read()!=HEADER): None
    while(ser.inWaiting()>0):
        in_byte = ser.read()
        if(in_byte!=FOOTER and i<message_size):
            message += in_byte
            i += 1
        elif(in_byte==FOOTER):
            return message

def get_data():
    message = ''
    data_list = []
    i = 0
    message = get_message()
    for j in range(6):
        if(message[i]==FINGER_ADDRESS[j]):
            data = ''
            i += 1
            for k in range(4):
                data += message[i]
                i += 1
            data = round(struct.unpack('f',data)[0],4)
            data_list.append(data)
    return data_list
